app/services/distribution_service.rb

```diff
@@ -0,0 +1,55 @@
+class DistributionService
+  attr_reader :error
+
+  def perform_distribution_service(&block)
+    distribution.transaction do
+      yield block
+    end
+  rescue ActiveRecord::RecordNotFound => e
+    Rails.logger.error "[!] #{self.class.name} failed to destroy distribution #{distribution_id} because it does not exist"
+    set_error(e)
+  rescue Errors::InsufficientAllotment => e
+    distribution.line_items.assign_insufficiency_errors(e.insufficient_items)
+    Rails.logger.error "[!] #{self.class.name} failed because of Insufficient Allotment #{distribution_organization.short_name}: #{distribution.errors.full_messages} [#{e.message}]"
+    set_error(e)
+  rescue StandardError => e
+    Rails.logger.error "[!] #{self.class.name} failed to destroy distribution for #{distribution_organization.short_name}: #{distribution.errors.full_messages} [#{e.inspect}]"
+    set_error(e)
+  ensure
+    return self
+  end
+
+  def success?
+    error.nil?
+  end
+
+  def distribution
+    # Return distribution if it has already been defined
+    return @distribution if @distribution
+
+    # Otherwise try to get this value with possibly
+    # provided distribution_id from initialize
+    if @distribution_id.present?
+      @distribution = Distribution.find(@distribution_id)
```

Comment:
This makes me nervous because we're not forcing the attribute to be defined from within the superclass. That said, if we move it private should be less of a risk.

Reply:
Ahh good point. I see that the `DistributionCreateService` has public access to the `distribution` and only that one.

```
  def create
    result = DistributionCreateService.new(distribution_params.merge(organization: current_organization), request_id).call

    if result.success?
      session[:created_distribution_id] = result.distribution.id
      redirect_to(distributions_path, notice: "Distribution created!") && return
    else
      @distribution = result.distribution
      flash[:error] = insufficient_error_message(result.error.message)
      # NOTE: Can we just do @distribution.line_items.build, regardless?
      @distribution.line_items.build if @distribution.line_items.count.zero?
      @items = current_organization.items.alphabetized
      @storage_locations = current_organization.storage_locations.alphabetized
      render :new
    end
```

Maybe a better approach is to keep that only in `DistributionCreateService`

Reply:
I agree about the tweaking the exposure of some of the methods -- could always use safe-nav or sprout a child method that calls the parent method?